### PR TITLE
video widget with selfie/new-front appearance

### PR DESCRIPTION
**Pull request by Thodoris1999**
The code is mostly based on `CaptureSelfieActivity` and [this](https://developer.android.com/guide/topics/media/camera.html#capture-video) and for new API [this](https://github.com/googlesamples/android-Camera2Video)
I added a `CameraUtils` to avoid duplicate code, however I have not changed `CaptureSelfieActivity` to use it in case it is a bad idea.

#### What has been done to verify that this works as intended? 
tested old api on 5.1 device, works and new api on 6.0 emulator, works

#### Why is this the best possible solution? Were any other approaches considered? 
I first considered to use the same activities/fragments and just add a boolean video flag for example so they capture video/images respectively, to avoid adding that many files and duplicating code, but I saw the suggestion of adding another activity get a like on this issue and it will probably make the code cleaner anyways. Also, I did not know where was the best place to store files so I just went with the ones in android's examples. I also tried pretty much any possible configuration of `MediaRecorder`, tested both using `TextureView` and `SurfaceView` on old API.

#### Are there any risks to merging this code? If so, what are they? 
It crashes on my Meizu m2 android 5.1

#### Do we need any specific form for testing your changes? If so, please attach one. 
A video widget with appearance set to "selfie" (I don't know how to attach xml files)

### CODE REVIEW

**Comment by Thodoris1999** [comment]
How do I un-fail that test on CircleCI?

**Comment by lognaturel** [comment]
Click on "Details" next to the failure. You'll see that a pmd check failed. You can run `./gradlew pmd` locally to see the report or you can see it by going to the artifacts tab and clicking on the pmd report.

**Comment by Thodoris1999** [comment]
When I run it locally I get:
Configuration 'androidTestCompile' in project ':collect_app' is deprecated. Use 'androidTestImplementation' instead.
Configuration 'compile' in project ':collect_app' is deprecated. Use 'implementation' instead.

FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':collect_app'.
> Failed to notify project evaluation listener.
   > Could not initialize class com.android.sdklib.repository.AndroidSdkHandler

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s

when I go to details it says "1 pmd violations found. See the report at: file:///home/circleci/work/collect_app/build/reports/pmd/pmd.html", but I can't find the artifacts tab or the pmd report. Can you elaborate on where I can find it?

**Comment by Thodoris1999** [comment]
The problem was that I didn't have a circleci account. Now it shows me the artifacts tab.

Also in the new api the camera preview doesn't take up the entire screen. This is not a bug, I just didn't adjust that yet to focus on the bug.

**Comment by lognaturel** [comment]
[selfie-video.xml.txt](https://github.com/opendatakit/collect/files/1798857/selfie-video.xml.txt) is a form that can be used. Adding `.txt.` is the best trick I've found for uploading forms.

Adding `CameraUtils` seems good to me. In some future when there is automated testing of this part of the code it might be better for those methods not to be static but this is a fine start.

> I first considered to use the same activities/fragments and just add a boolean video flag for example so they capture video/images respectively, to avoid adding that many files and duplicating code

How much code do you think actually could be shared between the two? I haven't looked at this in detail yet but at first glance it looks like keeping them separate is good.

Samsung Galaxy Tab SM-T285 - Android 5.1.1 and Motorola G 5th Gen XT1671 - Android 7.0 work well. What 5.1 device are you using?

Alcatel One Touch 5020D - Android 4.1.1 crashes with
```
java.lang.VerifyError: org/odk/collect/android/utilities/CameraUtils
	at org.odk.collect.android.widgets.VideoWidget.<init>(VideoWidget.java:132)
	at org.odk.collect.android.widgets.VideoWidget.<init>(VideoWidget.java:92)
	at org.odk.collect.android.widgets.WidgetFactory.createWidgetFromPrompt(WidgetFactory.java:173)
	at org.odk.collect.android.views.ODKView.<init>(ODKView.java:197)
	at org.odk.collect.android.activities.FormEntryActivity.createView(FormEntryActivity.java:1314)
	at org.odk.collect.android.activities.FormEntryActivity.showPreviousView(FormEntryActivity.java:1518)
	at org.odk.collect.android.activities.FormEntryActivity.onFling(FormEntryActivity.java:2849)
	at android.view.GestureDetector.onTouchEvent(GestureDetector.java:580)
	at org.odk.collect.android.activities.FormEntryActivity.dispatchTouchEvent(FormEntryActivity.java:1395)
```
I really don't know what could be causing that.

Some general things that will eventually need to be addressed (I understand this is a WIP but just want to get them out there so we don't lose track!):
- The sample code from the Android guides are great to use but the source should be clearly cited. There can be links to the guides or to [this](https://github.com/googlesamples/android-Camera2Basic/blob/master/Application/src/main/java/com/example/android/camera2basic/AutoFitTextureView.java) or [this](https://github.com/googlesamples/android-MediaRecorder/blob/master/Application/src/main/java/com/example/android/mediarecorder/MainActivity.java)
- User-facing strings will need to be in `strings.xml`

**Comment by Thodoris1999** [comment]
Sorry, I'm not familiar with automated testing, how can those methods be called from outside CameraUtils if they are not static?

The shareable code is most of the handling of the camera, which is not even the majority of the video capture class, now that I think of it keeping them separate and using CameraUtils probably is nicer.

> What 5.1 device are you using?

Meizu m2

> Alcatel One Touch 5020D - Android 4.1.1 crashes with

That's the first time I see anything like this... This is weird because all I did was move the existing VideoWidget code to CameraUtils. I can't find anything similar on the internet. Does the image selfie work before/after the changes? Did you test any other devices/images prior to Lollipop? I will make a stack overflow post and try to reproduce the problem

Thanks for reminding me these and for the links, I almost forgot. I didn't extract the string resources in case someone preferred to inform the user the video state with a button or something instead of toast.

Edit: https://stackoverflow.com/questions/49208104/verifyerror-while-checking-front-camera-prior-to-android-lollipop 
If that doesn't work, I'll move the front camera methods back where they were.

**Comment by Thodoris1999** [comment]
I'm done with most of this widget. I think the only think left is the bug on my device. I will provide more detail.

Everything is fine until recording is stopped and the app crashes. Here is the warning log from when I start to when I end the recording:
-11 18:58:55.314 18838-18838/org.odk.collect.android W/ContextImpl: Failed to ensure directory: /storage/sdcard1/Android/data/org.odk.collect.android/files
03-11 18:58:55.316 18838-18838/org.odk.collect.android E/MediaRecorder: MediaRecorder prepare   CallingPid =  18838  Callinguid=  10069
03-11 18:58:55.318 18838-18838/org.odk.collect.android E/MediaRecorder: java.lang.Throwable
                                                                            at android.media.MediaRecorder.prepare(MediaRecorder.java:792)
                                                                            at org.odk.collect.android.fragments.Camera2VideoFragment.setUpMediaRecorder(Camera2VideoFragment.java:579)
                                                                            at org.odk.collect.android.fragments.Camera2VideoFragment.startRecordingVideo(Camera2VideoFragment.java:594)
                                                                            at org.odk.collect.android.fragments.Camera2VideoFragment.onClick(Camera2VideoFragment.java:299)
                                                                            at android.view.View.performClick(View.java:4908)
                                                                            at android.view.View$PerformClick.run(View.java:20382)
                                                                            at android.os.Handler.handleCallback(Handler.java:815)
                                                                            at android.os.Handler.dispatchMessage(Handler.java:104)
                                                                            at android.os.Looper.loop(Looper.java:194)
                                                                            at android.app.ActivityThread.main(ActivityThread.java:5692)
                                                                            at java.lang.reflect.Method.invoke(Native Method)
                                                                            at java.lang.reflect.Method.invoke(Method.java:372)
                                                                            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:959)
                                                                            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:754)
03-11 18:58:55.343 265-19428/? E/MtkOmxVenc: [0xe8651c00] bufk = 0
03-11 18:58:55.343 265-19428/? E/VDO_LOG: drvb S
03-11 18:58:55.343 265-19428/? E/VDO_LOG: drvb E
03-11 18:58:55.343 265-19428/? E/VDO_LOG: drvb S
03-11 18:58:55.343 265-19428/? E/VDO_LOG: drvb E
03-11 18:58:55.344 265-19428/? E/MtkOmxVenc: [0xe8651c00] MtkOmxVenc::GetExtensionIndex Unknown parameter name: OMX.google.android.index.describeColorFormat
03-11 18:58:55.344 265-19428/? W/ACodec: do not know color format 0x7f000200 = 2130706944
03-11 18:58:55.344 265-19428/? E/MtkOmxVenc: [0xe8651c00] MtkOmxVenc::GetExtensionIndex Unknown parameter name: OMX.google.android.index.describeColorFormat
03-11 18:58:55.345 265-19428/? W/ACodec: do not know color format 0x7f000789 = 2130708361
03-11 18:58:55.345 265-19428/? E/MtkOmxVenc: [0xe8651c00] MtkOmxVenc::GetExtensionIndex Unknown parameter name: OMX.google.android.index.enableAndroidNativeBuffers
03-11 18:58:55.345 265-19428/? E/MtkOmxVenc: [0xe8651c00] MtkOmxVenc::GetExtensionIndex Unknown parameter name: OMX.google.android.index.describeColorFormat
03-11 18:58:55.345 265-19428/? W/ACodec: do not know color format 0x7f000789 = 2130708361
03-11 18:58:55.346 265-19428/? E/MtkOmxVenc: [0xe8651c00] MtkOmxVenc::GetExtensionIndex Unknown parameter name: OMX.google.android.index.storeGraphicBufferInMetaData
03-11 18:58:55.367 265-265/? W/StagefrightRecorder: Intended audio sample rate (8000) is too small and will be set to (12000)
03-11 18:58:55.368 265-265/? W/ServiceManager: Permission failure: android.permission.MODIFY_AUDIO_SETTINGS from uid=10069 pid=18838
                                               
                                               [ 03-11 18:58:55.368   265:  265 E/         ]
                                               Request requires android.permission.MODIFY_AUDIO_SETTINGS
03-11 18:58:55.373 265-265/? E/AudioALSAStreamIn: checkOpenStreamChannels(), wrong channels 0x10, use 0xc instead.
03-11 18:58:55.373 265-265/? E/AudioALSAStreamIn: checkOpenStreamSampleRate(), wrong sampleRate 16000, use 48000 instead.
03-11 18:58:55.373 265-265/? E/AudioALSAStreamManager: -openInputStream(), set fail, return NULL
03-11 18:58:55.391 265-19436/? E/OMXNodeInstance: getParameter(5c9:google.aac.encoder, ??(0x7f200406)) ERROR: UnsupportedIndex(0x8000101a)
03-11 18:58:55.514 721-1629/? E/Sensors: handleToDriver handle(0)
03-11 18:58:55.514 721-1629/? E/Sensors: handleToDriver handle(0)
03-11 18:58:55.522 721-1629/? E/Sensors: new setDelay handle(0),ns(66667000)m, error(0), index(2)
03-11 18:58:55.522 721-1629/? E/Accel: ACC batch: handle:0, en:0,samplingPeriodNs:66667000 maxBatchReportLatencyNs:0 
03-11 18:58:55.522 721-1629/? E/Sensors: sensor 0 go to common batch
03-11 18:58:55.523 721-2147/? E/Sensors: handleToDriver handle(3)
03-11 18:58:55.530 265-265/? W/MtkCam/CamClient/FDClient: (265)[stopFaceDetection] FD was not running
03-11 18:58:55.530 265-265/? W/MtkCam/CamClient/OTClient: (265)[stopOT] OT was not running
03-11 18:58:55.531 265-19414/? W/MtkCam/DisplayClient: [waitAndHandleReturnBuffers] vQueNode.empty()
03-11 18:58:55.531 265-19414/? W/MtkCam/MtkCamUtils: (CameraDisplay@ImgBufQue)[stopProcessor] intent to clear Done Que: size(6)!=0
03-11 18:58:55.534 721-1523/? E/libPerfService: Could not open '/d/ged/hal/custom_boost_gpu_freq'
03-11 18:58:55.534 721-1523/? E/libPerfService: error : 13, Permission denied
03-11 18:58:55.535 265-915/? W/MtkCam/Cam1Device: (915)(Default:1)[Cam1DeviceBase::initDisplayClient] NULL window is passed into...
03-11 18:58:55.535 265-915/? W/MtkCam/Cam1Device: (915)(Default:1)[Cam1DeviceBase::initDisplayClient] destroy the current display client(0xe87f90c0)...
03-11 18:58:55.570 18838-19410/org.odk.collect.android W/LegacyRequestMapper: convertRequestMetadata - control.awbRegions setting is not supported, ignoring value
03-11 18:58:55.570 18838-19410/org.odk.collect.android W/LegacyRequestMapper: Only received metering rectangles with weight 0.
03-11 18:58:55.586 18838-18838/org.odk.collect.android E/MediaRecorder: MediaRecorder start mz   CallingPid =  18838  Callinguid=  10069 mPath = /storage/emulated/0/Android/data/org.odk.collect.android/files/1520787535314.mp4
03-11 18:58:55.588 18838-18838/org.odk.collect.android E/MediaRecorder: java.lang.Throwable
                                                                            at android.media.MediaRecorder.start(MediaRecorder.java:851)
                                                                            at org.odk.collect.android.fragments.Camera2VideoFragment$4$1.run(Camera2VideoFragment.java:626)
                                                                            at android.os.Handler.handleCallback(Handler.java:815)
                                                                            at android.os.Handler.dispatchMessage(Handler.java:104)
                                                                            at android.os.Looper.loop(Looper.java:194)
                                                                            at android.app.ActivityThread.main(ActivityThread.java:5692)
                                                                            at java.lang.reflect.Method.invoke(Native Method)
                                                                            at java.lang.reflect.Method.invoke(Method.java:372)
                                                                            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:959)
                                                                            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:754)
03-11 18:58:55.591 265-265/? E/VideoQualityController: [init] not has video track
03-11 18:58:55.591 265-265/? W/MPEG4Writer: Disable low memory check by memfree due to profile not reach, videoFPS(0), videoWidth(0),  videoHeight(0)
03-11 18:58:55.601 265-19435/? W/AudioALSAStreamManager: -routingInputDevice(), input_device == current_input_device(0x80000004), return
03-11 18:58:55.618 265-19451/? W/MtkCam/ImgBuf: [IonHeap::onCreate] IonHeap@ Bad stride at 0-th plane: heapStridesInBytes(2432) -> (1297.066650)(pixels) -> StridesInBytes(2431)
03-11 18:58:55.623 265-19451/? W/MtkCam/ImgBuf: [IonHeap::onCreate] IonHeap@ Bad stride at 0-th plane: heapStridesInBytes(2432) -> (1297.066650)(pixels) -> StridesInBytes(2431)
03-11 18:58:55.625 265-19451/? W/MtkCam/ImgBuf: [IonHeap::onCreate] IonHeap@ Bad stride at 0-th plane: heapStridesInBytes(2432) -> (1297.066650)(pixels) -> StridesInBytes(2431)
03-11 18:58:55.628 265-19451/? W/MtkCam/ImgBuf: [IonHeap::onCreate] IonHeap@ Bad stride at 0-th plane: heapStridesInBytes(2432) -> (1297.066650)(pixels) -> StridesInBytes(2431)
03-11 18:58:55.631 265-19451/? W/MtkCam/ImgBuf: [IonHeap::onCreate] IonHeap@ Bad stride at 0-th plane: heapStridesInBytes(2432) -> (1297.066650)(pixels) -> StridesInBytes(2431)
03-11 18:58:55.633 265-19451/? W/MtkCam/ImgBuf: [IonHeap::onCreate] IonHeap@ Bad stride at 0-th plane: heapStridesInBytes(2432) -> (1297.066650)(pixels) -> StridesInBytes(2431)
03-11 18:58:55.637 265-19451/? W/MtkCam/ImgBuf: [IonHeap::onCreate] IonHeap@ Bad stride at 0-th plane: heapStridesInBytes(2432) -> (1297.066650)(pixels) -> StridesInBytes(2431)
03-11 18:58:55.640 265-19451/? W/MtkCam/ImgBuf: [IonHeap::onCreate] IonHeap@ Bad stride at 0-th plane: heapStridesInBytes(2432) -> (1297.066650)(pixels) -> StridesInBytes(2431)
03-11 18:58:55.755 265-9678/? E/ImgSensorDrv: [sendCommand]Err( 2181):[sendCommand] Err-ctrlCode (Success) 
03-11 18:58:55.755 265-9678/? E/MtkCam/HalSensor: [configure] CMD_SENSOR_GET_SENSOR_ROLLING_SHUTTER(10445)(2045) (configure){#1151:vendor/mediatek/proprietary/platform/mt6735/hardware/mtkcam/D1/hal/sensor/HalSensor.control.cpp}
03-11 18:58:55.775 721-1629/? E/Sensors: handleToDriver handle(0)
03-11 18:58:55.775 721-1629/? E/Sensors: handleToDriver handle(0)
03-11 18:58:55.778 721-1629/? E/Sensors: new setDelay handle(0),ns(10000000)m, error(0), index(2)
03-11 18:58:55.778 721-1629/? E/Accel: ACC batch: handle:0, en:0,samplingPeriodNs:10000000 maxBatchReportLatencyNs:0 
03-11 18:58:55.778 721-1629/? E/Sensors: sensor 0 go to common batch
03-11 18:58:55.778 721-1629/? E/Sensors: handleToDriver handle(0)
03-11 18:58:55.778 721-1629/? E/Sensors: go to batchsensor
03-11 18:58:55.778 721-2147/? E/Sensors: handleToDriver handle(0)
03-11 18:58:55.785 721-2147/? E/Sensors: new setDelay handle(0),ns(33000000)m, error(0), index(2)
03-11 18:58:55.785 721-1697/? E/Sensors: handleToDriver handle(3)
03-11 18:58:55.785 721-1697/? E/Sensors: handleToDriver handle(3)
03-11 18:58:55.785 721-1697/? E/Sensors: new setDelay handle(3),ns(10000000)m, error(-1), index(4)
03-11 18:58:55.785 721-1697/? E/Sensors: new setDelay handle(3),ns(10000000) err! go to hwmsen
03-11 18:58:55.785 721-1697/? E/GYRO: GYRO batch: handle:3, en:0,samplingPeriodNs:10000000, maxBatchReportLatencyNs:0 
03-11 18:58:55.785 721-1697/? E/Sensors: sensor 3 use old arch
03-11 18:58:55.785 721-1697/? E/Sensors: handleToDriver handle(3)
03-11 18:58:55.786 721-3775/? E/Sensors: handleToDriver handle(3)
03-11 18:58:55.786 721-3775/? E/Sensors: new setDelay handle(3),ns(33000000)m, error(-1), index(4)
03-11 18:58:55.786 721-3775/? E/Sensors: new setDelay handle(3),ns(33000000) err! go to hwmsen
03-11 18:58:55.874 721-1504/? E/WifiStateMachine:  ConnectedState (when=-3ms what=131155 arg1=433!CMD_RSSI_POLL 433 0 "Thom_D0048583" 80:c6:ab:48:47:fe rssi=-42 f=2412 sc=60 link=65 tx=0.2, 0.0, 0.0  rx=0.2 bcn=0 [on:0 tx:0 rx:0 period:3002] from screen [on:0 period:369113090] gl hn u24 rssi=-37 ag=0 hr ticks 0,0,0 ls-=0 [56,56,56,56,61] brc=0 lrc=0
03-11 18:58:55.875 721-1504/? E/WifiStateMachine:  L2ConnectedState (when=-4ms what=131155 arg1=433!CMD_RSSI_POLL 433 0 "Thom_D0048583" 80:c6:ab:48:47:fe rssi=-42 f=2412 sc=60 link=65 tx=0.2, 0.0, 0.0  rx=0.2 bcn=0 [on:0 tx:0 rx:0 period:1] from screen [on:0 period:369113091] gl hn u24 rssi=-37 ag=0 hr ticks 0,0,0 ls-=0 [56,56,56,56,61] brc=0 lrc=0
03-11 18:58:55.875 721-1504/? E/WifiStateMachine:  get link layer stats 0
03-11 18:58:55.877 721-1504/? E/WifiConfigStore: updateConfiguration freq=2412 BSSID=80:c6:ab:48:47:fe RSSI=-42 "Thom_D0048583"WPA_PSK
03-11 18:58:55.988 721-721/? E/WifiTrafficPoller: TRAFFIC_STATS_POLL true Token 2343 num clients 16
03-11 18:58:55.989 721-721/? E/WifiTrafficPoller:  packet count Tx=2619673 Rx=7381166
03-11 18:58:56.178 265-19402/? E/AeAlgo: [checkNightScene_v2p0()] Err:  1913:, NS_BrightTone_THDS < NS_OE_THDFLAT
03-11 18:58:56.206 18838-19411/org.odk.collect.android W/MALI: glDrawArrays:714: [MALI] glDrawArrays takes more than 5ms here. Total elapse time(us): 8429
03-11 18:58:56.279 18838-19131/org.odk.collect.android W/BufferQueueDump: [unnamed-18838-3] [void android::BufferQueueDump::addBuffer(const int&, const android::sp<android::GraphicBuffer>&, const android::sp<android::Fence>&, const int64_t&)] slot(2) acquired, seems to be abnormal, just update ...
03-11 18:58:56.287 18838-19411/org.odk.collect.android W/MALI: glDrawArrays:714: [MALI] glDrawArrays takes more than 5ms here. Total elapse time(us): 10956
03-11 18:58:56.357 18838-19411/org.odk.collect.android W/MALI: glDrawArrays:714: [MALI] glDrawArrays takes more than 5ms here. Total elapse time(us): 5469
03-11 18:58:56.418 265-19402/? E/AeAlgo: [checkNightScene_v2p0()] Err:  1913:, NS_BrightTone_THDS < NS_OE_THDFLAT
03-11 18:58:56.438 18838-19411/org.odk.collect.android W/MALI: glDrawArrays:714: [MALI] glDrawArrays takes more than 5ms here. Total elapse time(us): 5817
03-11 18:58:56.521 18838-19411/org.odk.collect.android W/MALI: glDrawArrays:714: [MALI] glDrawArrays takes more than 5ms here. Total elapse time(us): 7202
03-11 18:58:56.604 18838-19411/org.odk.collect.android W/MALI: glDrawArrays:714: [MALI] glDrawArrays takes more than 5ms here. Total elapse time(us): 10712
03-11 18:58:56.682 18838-19411/org.odk.collect.android W/MALI: glDrawArrays:714: [MALI] glDrawArrays takes more than 5ms here. Total elapse time(us): 6731
03-11 18:58:56.759 18838-19411/org.odk.collect.android W/MALI: glDrawArrays:714: [MALI] glDrawArrays takes more than 5ms here. Total elapse time(us): 5548
03-11 18:58:56.840 18838-19411/org.odk.collect.android W/MALI: glDrawArrays:714: [MALI] glDrawArrays takes more than 5ms here. Total elapse time(us): 7608
03-11 18:58:56.927 18838-19411/org.odk.collect.android W/MALI: glDrawArrays:714: [MALI] glDrawArrays takes more than 5ms here. Total elapse time(us): 11793
03-11 18:58:56.989 721-721/? E/WifiTrafficPoller: TRAFFIC_STATS_POLL true Token 2343 num clients 16
03-11 18:58:56.989 721-721/? E/WifiTrafficPoller:  packet count Tx=2619673 Rx=7381166
03-11 18:58:56.989 721-721/? E/WifiTrafficPoller: notifying of data activity 0
03-11 18:58:57.008 18838-19411/org.odk.collect.android W/MALI: glDrawArrays:714: [MALI] glDrawArrays takes more than 5ms here. Total elapse time(us): 12038
03-11 18:58:57.084 18838-19411/org.odk.collect.android W/MALI: glDrawArrays:714: [MALI] glDrawArrays takes more than 5ms here. Total elapse time(us): 5871
03-11 18:58:57.419 265-19430/? E/VDO_LOG: drvb S
03-11 18:58:57.419 265-19430/? E/VDO_LOG: drvb E
03-11 18:58:57.419 265-19430/? E/MtkOmxVenc: [0xe8651c00] VAL_CHIP_NAME_DENALI_1
03-11 18:58:57.419 265-19430/? E/MtkOmxVenc: [0xe8651c00] @@ [MtkOmxVenc::EncodeAVC] FrameWidth=1024, FrameHeight=768, BufWidth=1024, BufHeight=768
03-11 18:58:57.422 265-19430/? E/VDO_LOG: [eHalEMICtrlForRecordSize] VideoRecordSize_Set(BWC_SIZE(1024, 768))
                                          
                                          
                                          [ 03-11 18:58:57.422   265:19430 I/         ]
                                          [BWC INFO](19430): set_bwc_mm_property: propterty_id=2, value1=1024, value2=768 
03-11 18:58:57.424 265-19430/? E/VDO_LOG: [tid: 19430][WARN] the structure of h264_enc_prm is not sync!!
03-11 18:58:57.424 265-19430/? E/VDO_LOG: [tid: 19430] input frame rate:0x1e0000, I frame rate:30, bit rate:10000000, f:5
03-11 18:58:57.427 265-19430/? E/VDO_LOG: [tid: 19430] Don't Handle This kind of input!!
03-11 18:58:57.428 265-19444/? E/MPEG4Writer: The number of recorded samples is 0
03-11 18:58:57.428 265-19444/? W/MPEG4Writer: notify track err = -1007
03-11 18:58:57.428 265-19444/? W/MPEG4Writer: 0-duration samples found: 1
03-11 18:58:57.428 265-19444/? W/MPEG4Writer: 0-duration samples found: 1
03-11 18:58:57.429 265-19429/? E/MtkOmxVenc: [0xe8651c00] FlushInputPort -> mNumPendingInput(0)
03-11 18:58:57.431 265-19428/? W/GraphicBufferSource: Dropped back down to Loaded without Executing
03-11 18:58:57.472 265-20002/? E/BufferQueueProducer: [GraphicBufferSource](this:0xf1153400,id:56,api:1,p:18838,c:-1) dequeueBuffer: BufferQueue has been abandoned
03-11 18:58:57.473 18838-19411/org.odk.collect.android E/MALI: get_target_buffer:967: winsysp_window_buffer_get failed 12299
03-11 18:58:57.479 265-19435/? W/AudioALSAStreamManager: -routingInputDevice(), input_device == AUDIO_DEVICE_NONE(0x0), return
03-11 18:58:57.519 265-6009/? E/MPEG4Writer: MPEG4Writer::reset: !!!!!!ERROR during Track reset, do not write out moov, err=-1007
03-11 18:58:57.525 18838-18838/org.odk.collect.android E/MediaRecorder: stop failed: -1007
03-11 18:58:57.526 18838-18838/org.odk.collect.android E/UncaughtException: java.lang.RuntimeException: stop failed.
                                                                                at android.media.MediaRecorder.stop(Native Method)
                                                                                at org.odk.collect.android.fragments.Camera2VideoFragment.stopRecordingVideo(Camera2VideoFragment.java:657)
                                                                                at org.odk.collect.android.fragments.Camera2VideoFragment.onClick(Camera2VideoFragment.java:297)
                                                                                at android.view.View.performClick(View.java:4908)
                                                                                at android.view.View$PerformClick.run(View.java:20382)
                                                                                at android.os.Handler.handleCallback(Handler.java:815)
                                                                                at android.os.Handler.dispatchMessage(Handler.java:104)
                                                                                at android.os.Looper.loop(Looper.java:194)
                                                                                at android.app.ActivityThread.main(ActivityThread.java:5692)
                                                                                at java.lang.reflect.Method.invoke(Native Method)
                                                                                at java.lang.reflect.Method.invoke(Method.java:372)
                                                                                at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:959)
                                                                                at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:754)
03-11 18:58:57.549 18838-19411/org.odk.collect.android E/MALI: gles_state_set_error_internal:56: GLES ctx: 0x7f98406008, error code:0x505
03-11 18:58:57.572 18838-19411/org.odk.collect.android W/MALI: glDrawArrays:714: [MALI] glDrawArrays takes more than 5ms here. Total elapse time(us): 99769
03-11 18:58:57.572 18838-19411/org.odk.collect.android E/CameraDeviceGLThread-1: Received exception on GL render thread: 
                                                                                 java.lang.IllegalStateException: glDrawArrays: GLES20 error: 0x505
                                                                                     at android.hardware.camera2.legacy.SurfaceTextureRenderer.checkGlError(SurfaceTextureRenderer.java:537)
                                                                                     at android.hardware.camera2.legacy.SurfaceTextureRenderer.drawFrame(SurfaceTextureRenderer.java:346)
                                                                                     at android.hardware.camera2.legacy.SurfaceTextureRenderer.drawIntoSurfaces(SurfaceTextureRenderer.java:726)
                                                                                     at android.hardware.camera2.legacy.GLThreadManager$1.handleMessage(GLThreadManager.java:105)
                                                                                     at android.os.Handler.dispatchMessage(Handler.java:107)
                                                                                     at android.os.Looper.loop(Looper.java:194)
                                                                                     at android.os.HandlerThread.run(HandlerThread.java:61)
03-11 18:58:57.728 18838-18838/org.odk.collect.android E/AndroidRuntime: FATAL EXCEPTION: main
                                                                         Process: org.odk.collect.android, PID: 18838
                                                                         java.lang.RuntimeException: stop failed.
                                                                             at android.media.MediaRecorder.stop(Native Method)
                                                                             at org.odk.collect.android.fragments.Camera2VideoFragment.stopRecordingVideo(Camera2VideoFragment.java:657)
                                                                             at org.odk.collect.android.fragments.Camera2VideoFragment.onClick(Camera2VideoFragment.java:297)
                                                                             at android.view.View.performClick(View.java:4908)
                                                                             at android.view.View$PerformClick.run(View.java:20382)
                                                                             at android.os.Handler.handleCallback(Handler.java:815)
                                                                             at android.os.Handler.dispatchMessage(Handler.java:104)
                                                                             at android.os.Looper.loop(Looper.java:194)
                                                                             at android.app.ActivityThread.main(ActivityThread.java:5692)
                                                                             at java.lang.reflect.Method.invoke(Native Method)
                                                                             at java.lang.reflect.Method.invoke(Method.java:372)
                                                                             at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:959)
                                                                             at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:754)

**Comment by Thodoris1999** [comment]
What's wrong in the pmd check? I only changed 10 letters...

**Comment by Thodoris1999** [comment]
@lognaturel Can you please test pre-lollipop devices again? Testing on emulators is very inconsistent apparently

File: collect_app/src/main/java/org/odk/collect/android/views/CameraPreview.java [hunk 1]
```diff
@@ -56,8 +56,13 @@ public void surfaceCreated(SurfaceHolder holder) {
 
     public void surfaceDestroyed(SurfaceHolder holder) {
         if (camera != null) {
-            camera.stopPreview();
-            camera.release();
+            try {
+                camera.stopPreview();
+                camera.release();
+                Timber.w("Camera released");
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/views/CameraPreview.java]: This should be level `i`.

File: collect_app/src/main/java/org/odk/collect/android/widgets/VideoWidget.java [hunk 2]
```diff
@@ -92,6 +99,10 @@ public VideoWidget(Context context, FormEntryPrompt prompt, @NonNull FileUtil fi
         this.fileUtil = fileUtil;
         this.mediaUtil = mediaUtil;
 
+        String appearance = getFormEntryPrompt().getAppearanceHint();
+        selfie = appearance != null && appearance.equalsIgnoreCase("selfie");
+        Timber.d(Boolean.toString(selfie));
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/widgets/VideoWidget.java]: This one can be removed for production.

**Review by lognaturel — COMMENTED** [review]
👍It works very well on my Android 4.1.1 device.

Couple of small changes in line. All newly-added files will also need license headers. The ones that come from the Android project should have the headers copied from there. The ones you added should have a header as described in https://github.com/opendatakit/collect/blob/master/LICENSE.md#appendix-how-to-apply-the-apache-license-to-your-work with your name.

**Comment by lognaturel** [comment]
For your device, what seems to be happening is that when it tries to encode the video it runs into `StagefrightRecorder: Intended audio sample rate (8000) is too small and will be set to (12000)` and gets a permissions error.

**Comment by Thodoris1999** [comment]
@lognaturel I added `android.permission.MODIFY_AUDIO_SETTINGS` but it didn't fix the crash (it removed the log though). There is also this permission error: 
`03-11 18:58:55.534 721-1523/? E/libPerfService: error : 13, Permission denied`
I can't even find what libPerfService is...

**Comment by lognaturel** [comment]
Let's get it verified on a few more devices and see what we learn. If it works on most, I think we can start with this. The error you're getting looks like it may be specific to your hardware and it's ok if we don't support 100% of devices and Android versions for this feature.

**Comment by mmarciniak90** [comment]
I checked this changes on [selfie-video.xml.txt](https://github.com/opendatakit/collect/files/1841190/selfie-video.xml.txt) form.

Verified on Android: 4.1, 4.2, 4.4, 5.1, 6.0, 7.0, 8.0.

Android 4.1 does not have front camera so information about that appears:
screenshot_2018-03-23-11-33-04

I noticed problem with Android 4.2. I was not able to play video because file was not saved:
screenshot_2018-03-23-11-35-16

Beside this problem, everything works well on other Android versions.
@lognaturel should I create a separate issue for problem on Android 4.2 or @Thodoris1999 should fix this as part of this PR ?

**Comment by Thodoris1999** [comment]
I don't really know what causes this problem. Maybe using `Camera2VideoFragment::getVideoFilePath()` as the output file on the old api helps?

I think it is better for someone else to fix that issue. I can't test it on a real device and emulators are inconsistent in media recording for me. However, I do want to add one more cleanup commit in a couple hours (it won't affect the result).

**Comment by Thodoris1999** [comment]
By the way, I get this message when I run collect on Nexus S 4.2:
qemu-system-i386_2018-03-23_14-41-36
does anyone know what this means?

**Review by lognaturel — APPROVED** [review]

**Review by lognaturel — APPROVED** [review]

**Comment by lognaturel** [comment]
Thanks very much, @Thodoris1999, I know users will be very excited about this!

I'm not really worried about the Android 4.2 issues. It's not a very common version anymore. Let's not worry about it unless a user actually needs it. Then we can see what problem they are getting and try to address it. It looks like @mmarciniak90 and you are getting different issues.

**Comment by mmarciniak90** [comment]
I agree that it is a great feature!

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

**Pull request merged**